### DISCUSSION TITLE
Setting expiration of cache inside GetOrCreateAsync delegate

### DISCUSSION THREAD

**Discussion by zitzerd** [dotnet/aspnetcore#57021] [General]
With the default MemoryCache i use the CacheEntry expiration options to specify the expiration based on the result of the non cached operation, getting an jwt accesstoken for example.

With HybridCache i only see ways to sepcify expiration as an HybridCacheEntryOptions prior to calling the GetOrCreateAsync, is there a way to set it inside?

thanks

**Comment by engineering87** [comment]
The `AddHybridCache` method can be used to configure global defaults:

```csharp
builder.Services.AddHybridCache(options =>
    {
        options.MaximumPayloadBytes = 1024 * 1024;
        options.MaximumKeyLength = 1024;
        options.DefaultEntryOptions = new HybridCacheEntryOptions
        {
            Expiration = TimeSpan.FromMinutes(5), // set global expiration here
            LocalCacheExpiration = TimeSpan.FromMinutes(5)
        };
    });
```

**Reply by zitzerd** [reply]
Thanks for the reply. But what i meant was having different expirations defined to the getorset. As it's possible on the memorycache.

For example depending on a result of the operation (example getting a token) we define the ttl of the cache to prevent hitting the endpoint. 

This is possible with memory but i couldn't find the same possibility on hybrid.

Thanks

**Reply by engineering87** [reply]
The `GetOrCreateAsync` method can also accept a `HybridCacheEntryOptions` object to override the global defaults for a specific cache entry:

```csharp
public async Task<string> GetSomeInfoAsync(string name, int id, CancellationToken token = default)
    {
        var tags = new List<string> { "tag1", "tag2", "tag3" };
        var entryOptions = new HybridCacheEntryOptions
        {
            Expiration = TimeSpan.FromMinutes(1), // override expiration for current entry here
            LocalCacheExpiration = TimeSpan.FromMinutes(1)
        };
        return await _cache.GetOrCreateAsync(
            $"{name}-{id}", // Unique key to the cache entry
            async cancel => await GetDataFromTheSourceAsync(name, id, cancel),
            entryOptions,
            tags,
            token: token
        );
    }
```

**Reply by zitzerd** [reply]
Thanks again for the fast reply, maybe im missing something, probably am :P, but i don't see you setting the cache expiration inside te GetOrCreateAsync, im putting a sample of what i currently use with MemoryCache. Maybe it's not even a good way, but it serves the purpose 

Here is a basic example with MemoryCache
```C#
var accessToken = await _cache.GetOrCreateAsync(cacheKey, async (cacheEntry) =>
{
    //Do Some stuff that returns an expiration needed, for example an jwt token expiration
    var accessToken = GetAccessToken(uri);
    var jwt = new JwtSecurityTokenHandler().ReadJwtToken(accessToken.Token);
    //Specify my expiration inside the GetOrCreateAsync
    cacheEntry.AbsoluteExpiration = jwt.ValidTo;
    return  accessToken;
});

```    
Thanks Again!

**Comment by collinkruger** [comment]
Hi folks,
I'm looking into using HybridCache, and my team would find this as an important feature as well.

As an example, when a service is essentially acting as a reverse proxy for another backend, we want to be able to cache based off of the backend's `cache-control` response header.

Would it be possible to extend the HybridCache API to include a variant of `factory` that includes setting the HybridCacheEntryOptions for that cache entry/key.

**Comment by wilsonneto-swe** [comment]
Hey, folks!

@collinkruger and @zitzerd , currently, it is not possible and not part of the HybridCache design (yet); thanks for bringing these use cases.

There is already an issue regarding this use case request: [#56483 HybridCache: Enable setting expiration based on value in factory](https://github.com/dotnet/aspnetcore/issues/56483)

cc/ @mgravell

**Comment by mgravell** [comment]
The demand is noted. We can try to look at what we can add here.
